README.md
# serve-static-bun

[![npm version](https://img.shields.io/npm/v/serve-static-bun?style=flat-square)](https://npm.im/serve-static-bun)
[![npm bundle size](https://img.shields.io/bundlephobia/minzip/serve-static-bun?style=flat-square)](https://npm.im/serve-static-bun)
[![npm downloads](https://img.shields.io/npm/dt/serve-static-bun?style=flat-square)](https://npm.im/serve-static-bun)

Serve static files using `Bun.serve` or Bao.js.

Currently in beta. Aiming for similar features as [`expressjs/serve-static`](https://github.com/expressjs/serve-static).

## Install

This is a [Bun](https://bun.sh/) module available through the [npm registry](https://www.npmjs.com/). Installation is done using the [`bun install` command](https://github.com/oven-sh/bun#bun-install):

```sh
bun install serve-static-bun
```

## Usage

```js
import serveStatic from "serve-static-bun";
```

### `serveStatic(root, options)`

Create a new middleware function to serve files from within a given root directory. The file to serve will be determined by combining `req.url` with the provided root directory.

When a file is not found, it will send a 404 response. If a directory is accessed, but no index file is found, a 403 response will be sent.

When used in middleware mode, the 404 and 403 responses will not be sent and will instead return the context to Bao.js, so that other routes can continue.

By default, slashes are automatically collapsed in the path, and a trailing slash is added when the path is a directory. For example, if you have `blog/example/index.html` and access `https://example.com//blog///example`, it will redirect to `https://example.com/blog/example/`.

#### Options

##### `index`

By default this module will send "index.html" files in response to a request on a directory. To disable this, set it to `null`. To supply a new index, pass a string.

##### `dirTrailingSlash`

Redirect to trailing "/" when the pathname is a dir. Defaults to `true`.

##### `collapseSlashes`

Collapse all slashes in the pathname (`//blog///test` => `/blog/test`). Defaults to `true`.

##### `stripFromPathname`

Removes the first occurence of the specified string from the pathname. Is not defined by default (no stripping).

##### `headers`

Headers to add to the response. The "Content-Type" header cannot be overwritten. If you want to change the charset, use the `charset` option. If `collapseSlashes` or `dirTrailingSlash` is set, a "Location" header will be set if the pathname needs to be changed.

##### `dotfiles`

This option allows you to configure how the module handles dotfiles, i.e. files or directories that begin with a dot ("."). Dotfiles return a 403 by default (when this is set to "deny"), but this can be changed with this option.

##### `defaultMimeType`

The default mime type to send in the "Content-Type" HTTP header, when the file's cannot be determined. Defaults to `text/plain`.

##### `charset`

The "Content-Type" HTTP header charset parameter. Defaults to `utf-8`.

#### Middleware mode options

##### `middlewareMode`

When set to `"bao"`, it will return a Bao.js compatible handler function instead.

##### `handleErrors`

If set to `false`, in the case of a 403 or 404 response, the unmodified context will be returned to Bao.js. This allows you to handle the error yourself.  
If set to `true`, the error response will be sent to the client, without continuing the middleware chain.  
Defaults to `false`.

## Examples

### Serve files with vanilla `Bun.serve`

```ts
import serveStatic from "serve-static-bun";

Bun.serve({ fetch: serveStatic("public") });
```

### Serve files with Bao.js

```ts
import Bao from "baojs";
import serveStatic from "serve-static-bun";

const app = new Bao();

// *any can be anything
// We need to strip /assets from the pathname, because when the root gets combined with the pathname,
// it results in /assets/assets/file.js.
app.get(
	"/assets/*any",
	serveStatic("assets", { middlewareMode: "bao", stripFromPathname: "/assets" }),
);

app.get("/", (ctx) => ctx.sendText("Hello Bao!"));

app.listen();
```

### Serve only static files with Bao.js

**All** paths will be handled by `serve-static-bun`.

```ts
import Bao from "baojs";
import serveStatic from "serve-static-bun";

const app = new Bao();

// *any can be anything
app.get("/*any", serveStatic("web", { middlewareMode: "bao" }));

app.listen();
```

## License

[MIT](LICENSE)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/), and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.5.0] - 2022-12-29

### Added

- Descriptions for options in types.

### Changed

- Use typings for bun v0.4.0
- Updated descriptions for options in README.md
- Will now immediately send a 404 response if the file is not found, instead of redirecting to a normalized path before.
- `handleErrors` option is now `false` by default.

## [0.4.0] - 2022-11-21

### Changed

- Use typings for Bao.js v0.2.x

## [0.3.0] - 2022-10-31

### Changed

- The package is now an ES module instead of CJS.
- The `src/` folder is now included in the npm build.

## [0.2.0] - 2022-08-04

### Added

- [BREAKING] `dotfiles` option with `deny` as default. Your dotfiles will now respond with a 403 by default! No more leaking secrets.
- `defaultMimeType` option. Defaults to `text/plain`. Files that usually responded with `application/octet-stream` will now respond with the `defaultMimeType`, instead of automatically downloading, which was quite weird.

## [0.1.4] - 2022-07-12

### Changes

- Won't look for an index file if the path is not a directory.

## [0.1.3] - 2022-07-12

### Changes

- Use `Bun.file` instead of `node:fs`.

### Removed

- `mime` dependency. Now uses `Bun.file(file).type`!
- `fileEncoding` option is gone, replaced by `charset` only.

## [0.1.1] - 2022-07-12

### Removed

- Source maps

## [0.1.0] - 2022-07-12

### Added

- Initial release
- Support for `Bun.serve`
- Support for `Bao.js`


src/utils/get-file-info.ts
import type { FileBlob } from "bun";
import { isErrorlike } from "../types";

export interface FileInfo {
	/**
	 * A blob with the file's info.
	 *
	 * @see Bun.FileBlob
	 */
	blob: FileBlob;

	/**
	 * Whether the file exists.
	 */
	exists: boolean;

	/**
	 * Whether the file is a file. If `false`, it is a directory.
	 */
	isFile: boolean;

	/**
	 * The mime type of the file, if it can be determined.
	 * If it cannot be determined, it will be `undefined`.
	 */
	mimeType?: string;
}

function getMimeType({ type }: FileBlob) {
	const charsetIndex = type.indexOf(";charset");
	return charsetIndex !== -1 ? type.substring(0, charsetIndex) : type;
}

/**
 * Returns information about a file.
 *
 * @param path The path to the file
 * @returns Information about the file
 */
export default async function getFileInfo(path: string) {
	const info: FileInfo = {
		blob: Bun.file(path),
		exists: false,
		isFile: false,
	};

	try {
		await info.blob.arrayBuffer();
		info.exists = true;
		info.isFile = true;
		const mimeType = getMimeType(info.blob);
		info.mimeType = mimeType === "application/octet-stream" ? undefined : mimeType;
	} catch (error) {
		if (isErrorlike(error)) {
			switch (error.code) {
				case "EISDIR":
					info.exists = true;
					break;
			}
		}
	}

	return info;
}


src/serve-static.ts
import type { Context } from "baojs/dist/context";
import getBaoMiddleware from "./middleware/bao";
import getFileInfo, { type FileInfo } from "./utils/get-file-info";

/**
 * Options for serveStatic().
 */
interface ServeStaticBaseOptions {
	/**
	 * By default this module will send "index.html" files in response to a request on a directory.
	 * To disable this, set it to `null`. To supply a new index, pass a string.
	 *
	 * @default "index.html"
	 */
	index?: string | null;

	/**
	 * Redirect to trailing "/" when the pathname is a dir.
	 *
	 * @default true
	 */
	dirTrailingSlash?: boolean;

	/**
	 * Collapse all slashes in the pathname (`//blog///test` => `/blog/test`).
	 *
	 * @default true
	 */
	collapseSlashes?: boolean;

	/**
	 * Remove the first occurence of the specified string from the pathname.
	 * Is not defined by default (no stripping).
	 */
	stripFromPathname?: string;

	/**
	 * Headers to add to the response. The "Content-Type" header cannot be overwritten. If you want to
	 * change the charset, use the `charset` option. If `collapseSlashes` or `dirTrailingSlash` is set,
	 * a "Location" header will be set if the pathname needs to be changed.
	 */
	headers?: HeadersInit;

	/**
	 * This option allows you to configure how the module handles dotfiles, i.e. files or directories that begin with a dot (".").
	 * Dotfiles return a 403 by default (when this is set to "deny"), but this can be changed with this option.
	 *
	 * @default "deny"
	 */
	dotfiles?: "allow" | "deny";

	/**
	 * The default mime type to send in the "Content-Type" HTTP header, when the file's cannot be determined.
	 *
	 * @default "text/plain"
	 */
	defaultMimeType?: string;

	/**
	 * The "Content-Type" HTTP header charset parameter.
	 *
	 * @default "utf-8"
	 */
	charset?: string;
}

/**
 * Options for serveStatic() when used as a middleware.
 */
interface ServeStaticMiddlewareOptions extends ServeStaticBaseOptions {
	/**
	 * The type of middleware to generate.
	 * When set to `"bao"`, it will return a Bao.js compatible handler function instead.
	 */
	middlewareMode: "bao";

	/**
	 * If set to `false`, in the case of a 403 or 404 response, the unmodified context will be returned to Bao.js.
	 * This allows you to handle the error yourself.
	 *
	 * If set to `true`, the error response will be sent to the client, without continuing the middleware chain.
	 *
	 * @default false
	 */
	handleErrors?: boolean;
}

/**
 * Type guard for ServeStaticMiddlewareOptions.
 */
function isMiddleware(options: ServeStaticOptions): options is ServeStaticMiddlewareOptions {
	return Object.hasOwn(options, "middlewareMode");
}

type ServeStaticOptions = ServeStaticBaseOptions | ServeStaticMiddlewareOptions;

/**
 * Get the correct pathname from the requested URL.
 * @param url The requested URL
 * @param stripFromPathname The string to remove from the pathname, if necessary
 */
function getPathname(
	{ pathname }: URL,
	stripFromPathname: ServeStaticBaseOptions["stripFromPathname"],
) {
	return stripFromPathname ? pathname.replace(stripFromPathname, "") : pathname;
}

/**
 * Get the normalized path to redirect to.
 * @param pathname The requested pathname
 * @param requestedFile The requested file
 * @param options The serveStatic() options
 */
async function getRedirectPath(
	pathname: string,
	{ isFile }: FileInfo,
	{
		collapseSlashes,
		dirTrailingSlash,
	}: Pick<ServeStaticBaseOptions, "collapseSlashes" | "dirTrailingSlash">,
) {
	let redirectPath = pathname;

	// Normalize slashes
	if (collapseSlashes) {
		const pkg = await import("./utils/collapse-slashes");
		redirectPath = pkg.collapseSlashes(redirectPath, {
			keepTrailing: redirectPath.endsWith("/"), // Preserve trailing slash if it exists
		});
	}

	// Add trailing slash
	if (dirTrailingSlash && !isFile && !redirectPath.endsWith("/")) {
		redirectPath = `${redirectPath}/`;
	}

	return redirectPath;
}

/**
 * Get the file to serve, either the requested file or the folder's index file.
 * @param pathname The requested pathname
 * @param requestedFile The requested file
 * @param root The root path
 * @param options The serveStatic() options
 * @returns The file to serve, or null if none exists
 */
async function getFileToServe(
	pathname: string,
	requestedFile: FileInfo,
	root: string,
	{ index, dotfiles }: Pick<ServeStaticBaseOptions, "index" | "dotfiles">,
) {
	const isDotfile = pathname.split("/").pop()?.startsWith(".");
	if (requestedFile.isFile && (!isDotfile || dotfiles === "allow")) {
		return requestedFile;
	}

	// If it is a folder and it has an index
	const indexFile = index === null ? null : await getFileInfo(`${root}/${pathname}/${index}`);
	if (indexFile?.exists && indexFile.isFile) {
		return indexFile;
	}

	return null;
}

/**
 * For use with {@link Bun.serve}'s fetch function directly.
 *
 * @example
 *
 * ```ts
 * import serveStatic from "serve-static-bun";
 *
 * Bun.serve({ fetch: serveStatic("frontend") });
 * ```
 * @param root The path to the static files to serve
 * @param options
 */
export default function serveStatic(
	root: string,
	options?: ServeStaticBaseOptions,
): (req: Request) => Promise<Response>;

/**
 * For use as a bao middleware.
 *
 * @example
 *
 * Serve files with Bao.js
 *
 * ```ts
 * import Bao from "baojs";
 * import serveStatic from "serve-static-bun";
 *
 * const app = new Bao();
 *
 * // *any can be anything
 * // We need to strip /assets from the pathname, because when the root gets combined with the pathname,
 * // it results in /assets/assets/file.js.
 * app.get("/assets/*any", serveStatic("assets", { middlewareMode: "bao", stripFromPathname: "/assets" }));
 *
 * app.get("/", (ctx) => ctx.sendText("Hello Bao!"));
 *
 * app.listen();
 * ```
 *
 * @example
 *
 * Serve only static files with Bao.js
 *
 * **All** paths will be handled by `serve-static-bun`.
 *
 * ```ts
 * import Bao from "baojs";
 * import serveStatic from "serve-static-bun";
 *
 * const app = new Bao();
 *
 * // *any can be anything
 * app.get("/*any", serveStatic("web", { middlewareMode: "bao" }));
 *
 * app.listen();
 * ```
 * @param root The path to the static files to serve
 * @param options
 */
export default function serveStatic(
	root: string,
	options: ServeStaticMiddlewareOptions,
): (ctx: Context) => Promise<Context>;

export default function serveStatic(root: string, options: ServeStaticOptions = {}) {
	root = `${process.cwd()}/${root}`;
	const {
		index = "index.html",
		dirTrailingSlash = true,
		collapseSlashes = true,
		stripFromPathname,
		headers,
		dotfiles = "deny",
		defaultMimeType = "text/plain",
		charset = "utf-8",
	} = options;
	const wantsMiddleware = isMiddleware(options);

	const getResponse = async (req: Request) => {
		const pathname = getPathname(new URL(req.url), stripFromPathname);
		const requestedFile = await getFileInfo(`${root}/${pathname}`);

		// If path does not exists, return 404
		if (!requestedFile.exists) {
			return new Response("404 Not Found", {
				status: 404,
				headers: {
					...headers,
					"Content-Type": `text/plain; charset=${charset}`,
				},
			});
		}

		// Redirect to normalized path, if needed
		const redirectPath = await getRedirectPath(pathname, requestedFile, {
			collapseSlashes,
			dirTrailingSlash,
		});
		if (redirectPath !== pathname) {
			return new Response(undefined, {
				status: 308, // Permanent Redirect, cacheable
				headers: {
					...headers,
					Location: redirectPath,
				},
			});
		}

		// Serve file or index, if one of them exists
		const fileToServe = await getFileToServe(pathname, requestedFile, root, { index, dotfiles });
		if (fileToServe) {
			return new Response(fileToServe.blob, {
				headers: {
					...headers,
					"Content-Type": `${fileToServe.mimeType ?? defaultMimeType}; charset=${charset}`,
				},
			});
		}

		// Fallback to 403
		return new Response("403 Forbidden", {
			status: 403,
			headers: {
				...headers,
				"Content-Type": `text/plain; charset=${charset}`,
			},
		});
	};

	if (wantsMiddleware) {
		const { middlewareMode, handleErrors = false } = options;

		switch (middlewareMode) {
			case "bao":
				return getBaoMiddleware(getResponse, handleErrors);

			// No default
		}
	}

	return getResponse;
}
